File: Week_08/kotlin/src/_1030_MatrixCellsInDistanceOrder_.kt
import java.io.BufferedInputStream
import java.io.FileInputStream
import java.util.*
import kotlin.Comparator

class _1030_ {
    class Solution {
        fun allCellsDistOrder(R: Int, C: Int, r0: Int, c0: Int): Array<IntArray> {
            val ret = Array<IntArray>(R * C) { IntArray(2) }
            val inputStream = BufferedInputStream(FileInputStream(""))
            inputStream.use { }
            for (i in 0 until R) {
                for (j in 0 until C) {
                    ret[i * C + j][0] = i
                    ret[i * C + j][1] = j
                }
            }
            ret.sortWith(Comparator { l, r -> Math.abs(l[0] - r0) + Math.abs(l[1] - c0) - Math.abs(r[0] - r0) - Math.abs(r[1] - c0) })
            return ret
        }
    }

    class UsingMath {
        fun allCellsDistOrder(R: Int, C: Int, r0: Int, c0: Int): Array<IntArray> {
            val maxDist = maxOf(r0, R - 1 - r0) + maxOf(c0, C - 1 - c0)
            val ret = ArrayList<IntArray>(R * C) .apply{ add(intArrayOf(r0, c0)) }
            val dirs = arrayOf(intArrayOf(1, 1), intArrayOf(1, -1), intArrayOf(-1, -1), intArrayOf(-1, 1))
            var r = r0
            var c = c0
            for (dist in 1..maxDist) {
                r--
                for (i in 0..3) {
                    while (((i and 1) == 0 && r != r0) || ((i and 1) != 0 && c != c0)) {
                        if (r in 0 until R && c in 0 until C) {
                            ret += intArrayOf(r, c)
                        }
                        r += dirs[i][0]
                        c += dirs[i][1]
                    }
                }
            }
            return ret.toTypedArray()
        }
    }

    class UsingBucketSort {
        fun allCellsDistOrder(R: Int, C: Int, r0: Int, c0: Int): Array<IntArray> {
            fun dist(r: Int, c : Int) = Math.abs(r - r0) + Math.abs(c - c0)
            val maxDist = maxOf(r0, R - 1 - r0) + maxOf(c0, C - 1 - c0)
            val buckets = Array(maxDist + 1) { LinkedList<IntArray>() }
            val ret = ArrayList<IntArray>(R * C)
            for (i in 0 until R) {
                for (j in 0 until C) {
                    buckets[dist(i, j)].offer(intArrayOf(i, j))
                }
            }
            buckets.forEach { ret.addAll(it) }
            return ret.toTypedArray()
        }
    }
}

File: Week_08/cpp/src/_1030_MatrixCellsInDistanceOrder.cpp
#include <vector>
using namespace std;
class UsingMath
{
public:
    // 几何法，尚未搞懂点数是怎么计算的
    // O((R + C) ^ 2) O(1)
    vector<vector<int>> allCellsDistOrder(int R, int C, int r0, int c0)
    {
        vector<vector<int>> dirs{{1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
        vector<vector<int>> ret{{r0, c0}};
        int r = r0, c = c0;
        int maxDist = max(r0, R - r0 - 1) + max(c0, C - c0 - 1);
        for (int dist = 1; dist <= maxDist; dist++)
        {
            r--;
            for (int i = 0; i < 4; i++)
            {
                while ((!(i & 1) && r != r0) || ((i & 1) && c != c0))
                {
                    if (r >= 0 && r < R && c >= 0 && c < C)
                    {
                        ret.push_back({r, c});
                    }
                    r += dirs[i][0];
                    c += dirs[i][1];
                }
            }
        }
        return ret;
    }
};
class UsingBucketSort
{
public:
    vector<vector<int>> allCellsDistOrder(int R, int C, int r0, int c0)
    {
        int maxDist = max(r0, R - r0 - 1) + max(c0, C - c0 - 1);
        vector<vector<vector<int>>> buckets(maxDist + 1);

        const auto dist = [&](int r, int c) -> int { return abs(r - r0) + abs(c - c0); };
        for (int i = 0; i < R; i++)
        {
            for (int j = 0; j < C; j++)
            {
                // vector<int> t = {i, j};
                // buckets[dist(i, j)].push_back(move(t));
                buckets[dist(i, j)].push_back({i, j});
            }
        }
        vector<vector<int>> ret;
        for (const auto &bucket : buckets)
        {
            for (const auto &it : bucket)
            {
                ret.push_back(it);
            }
        }
        return ret;
    }
};